### PR TITLE
Fix PythonOp with input which has no gradient

### PR DESCRIPTION
**Pull request by wschin**
As title. We shouldn't generate gradient for tensors which don't require gradient. Otherwise, when a tensor is consumed by two op, backward graph may have two gradient tensors with the same name. A new test is added. Without our fix, that test may fail.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc [hunk 1]
```diff
@@ -258,11 +258,11 @@ void PythonOpGradBase::RunBackward(OpKernelContext* context,
 
 void PythonOpGradBase::SetOutputs(OpKernelContext* context, std::vector<OrtValue>& returned_ortvalues) const {
   auto* ctx_internal = reinterpret_cast<onnxruntime::OpKernelContextInternal*>(context);
-  auto outputs_count = static_cast<size_t>(ctx_internal->OutputCount());
-  // It's possible that Pytorch returns None as gradient and ORT Python side may skip them.
-  // In that case, returned_args may contain less arguments.
-  outputs_count = outputs_count > returned_ortvalues.size() ? returned_ortvalues.size() : outputs_count;
+  const auto outputs_count = static_cast<size_t>(ctx_internal->OutputCount());
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:261]: `None`s are filtered out in backward runner (written in Python) so the previous comment is not needed anymore.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py [hunk 2]
```diff
@@ -660,3 +660,53 @@ def get_inner_module_call_result(x, device, use_ort):
...
+                output_size,
+                device=torch.cuda.current_device(),
+                dtype=torch.float))
+
+            # Always initialize bias to zero.
+            with torch.no_grad():
+                self.bias.uniform_()
+
+        def forward(self, x):
+            a, b = self.fun(x, self.bias)
```
Comment (wschin) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:696]: Tensor `x` are `shared` by two `PythonOp`.

Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:696]: As we synced offline, we'd better try a case x.requires_grad=True.

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc (hunk 1 above)
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:261]: seems our cases all assume None happens at the end. E.g. value1, value2, None, None

maybe we should add cases for forward/backward, having outputs like value1, None, value2, None, None, ...,value3.
what do you think?

Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:261]: Yep. It's already a part of my local changes you saw today.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py (hunk 2 above)
Comment (wschin) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:696]: Done.

**Review by tlh20 — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py [hunk 3]
```diff
@@ -660,3 +660,58 @@ def get_inner_module_call_result(x, device, use_ort):
...
+            super(TwoOutputModel, self).__init__()
+            self.fun = TwoOutputFunction.apply
+            self.bias = Parameter(torch.empty(
+                output_size,
+                device=torch.cuda.current_device(),
+                dtype=torch.float))
+
+            # Always initialize bias to zero.
+            with torch.no_grad():
+                self.bias.uniform_()
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:693]: Doesn't this initialize `self.bias` to uniform random numbers between 0 and 1?

Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:693]: Yeah so the comment is wrong. :) thanks!

Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py]: Fixed

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:261]: I will add a test seperately.

**Review by mrry — APPROVED** [review]
Looks good to me, thanks!

**Comment by pengwa** [comment]
Thanks a lot! @mrry.
